lib/Sema/TypeCheckAttr.cpp

```diff
@@ -1217,33 +1225,87 @@ visitDynamicMemberLookupAttr(DynamicMemberLookupAttr *attr) {
   // This attribute is only allowed on nominal types.
   auto decl = cast<NominalTypeDecl>(D);
   auto type = decl->getDeclaredType();
-  
-  // Look up `subscript(dynamicMember:)` candidates.
-  auto subscriptName = DeclName(TC.Context, DeclBaseName::createSubscript(),
-                                TC.Context.Id_dynamicMember);
-  auto candidates = TC.lookupMember(decl, type, subscriptName);
-  
-  // If there are no candidates, then the attribute is invalid.
-  if (candidates.empty()) {
+  auto &ctx = decl->getASTContext();
+
+  auto emitInvalidTypeDiagnostic = [&]() {
     TC.diagnose(attr->getLocation(), diag::invalid_dynamic_member_lookup_type,
                 type);
     attr->setInvalid();
+  };
+
+  // Look up all subscript candidates
+  auto candidates =
+      TC.lookupMember(decl, type, DeclBaseName::createSubscript());
+
+  if (candidates.empty()) {
+    emitInvalidTypeDiagnostic();
     return;
   }
 
-  // If no candidates are valid, then reject one.
-  auto oneCandidate = candidates.front();
-  candidates.filter([&](LookupResultEntry entry, bool isOuter) -> bool {
+  // We now have all the subscripts that the user declared inside the type.
+  //
+  // 1. A `subscript(dynamicMember <argument label>:)` is acceptable.
+  //    We need at least one candidate to satisfy the @dynamicMemberLookup
+  //    requirement.
+  // 2. A `subscript(dynamicMember <no argument label>:)` should be diagnosed.
+  // 3. A `subscript` without `dynamicMember` should be diagnosed.
+  //
+  // Let's validate the subscript candidates while ignoring the labels.
+  candidates.filter([&](const LookupResultEntry entry, bool isOuter) {
```

Comment:
Sorry, I think we'd better do a minimum amount of lookup to cover a happy case, so:

- Try lookup with correct label
- If that turns up no candidates:
  - Emit an error
  - Lookup with base name only
  - filter candidates
  - emit a note for each candidate which matches on everything besides a label

Reply:
Okay, I have revised the implementation :)